src/remote_write/tests/equivalence_test.rs

```diff
@@ -0,0 +1,177 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This test is used to verify the correctness of the pooled parser rigorously
+//! (50 iterations, 25 iterations for each input data), by comparing the results
+//! with the prost parser.
+//!
+//! Test with `--features unsafe-split` to enable the unsafe optimization.
+
+use std::{fs, sync::Arc};
+
+use bytes::Bytes;
+use pb_types::{Exemplar, Label, MetricMetadata, Sample, TimeSeries, WriteRequest};
+use prost::Message;
+use remote_write::pooled_parser::PooledParser;
+use tokio::task::JoinHandle;
+
+const ITERATIONS: usize = 50;
+
+fn load_test_data() -> (Bytes, Bytes) {
+    let data1 = Bytes::from(
+        fs::read("tests/workloads/1709380533560664458.data").expect("test data load failed"),
+    );
+
+    let data2 = Bytes::from(
+        fs::read("tests/workloads/1709380533705807779.data").expect("test data load failed"),
+    );
+
+    (data1, data2)
+}
+
+fn parse_with_prost(data: &Bytes) -> WriteRequest {
+    WriteRequest::decode(data.clone()).expect("prost decode failed")
+}
+
+async fn parse_with_pooled(data: &Bytes) -> WriteRequest {
+    let data_copy = data.clone();
+    let parser = PooledParser;
+    let pooled_request = parser
+        .decode_async(data_copy)
+        .await
+        .expect("pooled decode failed");
+
+    // Convert pooled types to pb_types to compare with prost.
+    let mut write_request = WriteRequest {
+        timeseries: Vec::new(),
+        metadata: Vec::new(),
+    };
+
+    for pooled_ts in pooled_request.timeseries.iter() {
```

Comment:
```suggestion
    for pooled_ts in &pooled_request.timeseries {
```

The same changes apply to for loops below also.